Tests/MapboxNavigationTests/VanishingRouteLineTests.swift

```diff
@@ -326,4 +326,72 @@ class VanishingRouteLineTests: TestCase {
         XCTAssertTrue(expectedFractionTraveled == navigationMapView.fractionTraveled, "Failed to stop updating fractionTraveled when user off the route line.")
     }
     
+    func testSwitchshowsRestrictedAreasOnRoute() {
+        let routeProgress = getRouteProgress()
+        let route = routeProgress.route
+        let coordinate = route.shape!.coordinates[1]
+
+
+        navigationMapView.trafficUnknownColor = UIColor.blue
+        navigationMapView.trafficModerateColor = UIColor.red
+        navigationMapView.traversedRouteColor = UIColor.clear
+        setUpCameraZoom(at: 16.0)
+        
+        navigationMapView.routeLineTracksTraversal = true
+        navigationMapView.show([route], legIndex: 0)
+        navigationMapView.addArrow(route: route, legIndex: 0, stepIndex: 2)
+        navigationMapView.updateUpcomingRoutePointIndex(routeProgress: routeProgress)
+        navigationMapView.travelAlongRouteLine(to: coordinate)
+
+        let fractionTraveled = navigationMapView.fractionTraveled
+        let fractionTraveledNextDown = Double(CGFloat(fractionTraveled).nextDown)
+        let expectedExpressionString = "[step, [line-progress], [rgba, 0.0, 0.0, 255.0, 1.0], 0.0, [rgba, 0.0, 0.0, 0.0, 0.0], \(fractionTraveledNextDown), [rgba, 0.0, 0.0, 0.0, 0.0], \(fractionTraveled), [rgba, 0.0, 0.0, 255.0, 1.0], 0.9425498181625797, [rgba, 0.0, 0.0, 255.0, 1.0], 0.9425498181625799, [rgba, 255.0, 0.0, 0.0, 1.0]]"
+
+        let layerIdentifier = route.identifier(.route(isMainRoute: true))
+        
+        var allLayerIds = navigationMapView.mapView.mapboxMap.style.allLayerIdentifiers.map({ $0.id })
+        guard let indexOfMainRouteLayer = allLayerIds.firstIndex(of: route.identifier(.route(isMainRoute: true))),
+              let indexOfArrowStrokeLayer = allLayerIds.firstIndex(of: NavigationMapView.LayerIdentifier.arrowStrokeLayer) else {
+                  XCTFail("Failed to find all the layers")
+                  return
+              }
+        XCTAssert(indexOfMainRouteLayer < indexOfArrowStrokeLayer, "Arrow stroke layer should be above main route layer")
+        
+        // When the `NavigationMapView.showsRestrictedAreasOnRoute` is turned on during active navigation with `routeLineTracksTraversal` enabled,
+        // the previous vanishing effect should be kept, and the new restricted areas layer should be added
+        // above the main route line layer but below the arrow stroke layer.
+        navigationMapView.showsRestrictedAreasOnRoute = true
+        do {
+            let layer = try navigationMapView.mapView.mapboxMap.style.layer(withId: layerIdentifier) as! LineLayer
```

Comment:
NIT: In general it's better to check if object is valid, instead doing of force-unwrapping, this saves time when running a bunch of tests locally, because crash during force unwrapping will stop execution.

Reply:
Thanks for the suggestion, the force unwrapping will be fixed in #3687 .